pkg/registry/resource/utils.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/kubernetes/pkg/apis/resource"
+)
+
+// AuthorizedForAdmin checks if the request is authorized to get admin access to devices
+// based on namespace label
+func AuthorizedForAdmin(ctx context.Context, deviceRequests []resource.DeviceRequest, namespaceName string, nsClient v1.NamespaceInterface) field.ErrorList {
+	var allErrs field.ErrorList
+	adminRequested := false
+	var adminAccessPath *field.Path
+
+	// no need to check old request since spec is immutable
+
+	for i := range deviceRequests {
+		value := deviceRequests[i].AdminAccess
+		if value != nil && *value {
+			adminRequested = true
+			adminAccessPath = field.NewPath("spec", "devices", "requests").Index(i).Child("adminAccess")
+			break
+		}
+	}
+	if !adminRequested {
+		// No need to validate unless admin access is requested
+		return allErrs
+	}
+
+	// Retrieve the namespace object from the store
+	ns, err := nsClient.Get(ctx, namespaceName, metav1.GetOptions{})
+	if err != nil {
+		return append(allErrs, field.InternalError(adminAccessPath, fmt.Errorf("could not retrieve namespace to verify admin access: %w", err)))
+	}
+	if ns.Labels[resource.DRAAdminNamespaceLabelKey] != "true" {
+		return append(allErrs, field.Forbidden(adminAccessPath, fmt.Sprintf("admin access to devices requires the `%s: true` label on the containing namespace", resource.DRAAdminNamespaceLabelKey)))
+	}
+
+	return allErrs
+}
+
+// AuthorizedForAdminStatus checks if the request status is authorized to get admin access to devices
+// based on namespace label
+func AuthorizedForAdminStatus(ctx context.Context, newAllocationResult, oldAllocationResult []resource.DeviceRequestAllocationResult, namespaceName string, nsClient v1.NamespaceInterface) field.ErrorList {
+	var allErrs field.ErrorList
+	var adminAccessPath *field.Path
+
+	if wasGranted, _ := adminRequested(oldAllocationResult); wasGranted {
+		if wasGranted {
```

Comment:
drop the extra nested `if wasGranted` check

then go ahead and squash to a single commit and lgtm

Reply:
done